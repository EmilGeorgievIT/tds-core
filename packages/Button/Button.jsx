import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { componentWithName, or } from '@tds/util-prop-types'

import { warn } from '../../shared/utils/warn'
import safeRest from '../../shared/utils/safeRest'
import media from '../../shared/utils/media/media'

const preventDisabling = ({ disabled, ...props }) => {
  if (disabled) {
    warn('Button', 'Buttons are not able to be disabled.')
  }

  return props
}

const BaseButton = styled.button`
  border-width: 0;
  border-radius: 4px;

  height: 3.25rem;
  margin: 0;
  padding: 0 2rem;

  cursor: pointer;

  background: none;
  transition: background 0.2s;

  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;

  word-wrap: break-word;

  font-size: 1rem;
  letter-spacing: -0.8px;
  line-height: 1.5;
  font-weight: 700;


  ${media.greaterThan('sm')`
    display: inline-flex;

    width: auto;
    min-width: 180px;
  `}
`

const ButtonPrimary = styled(BaseButton)`
  background-color: #248700;
  color: #FFFFFF;
`

const ButtonSecondary = styled(BaseButton)`
  background-color: #4b286d;
  color: #FFFFFF;
`

const ButtonInverted = styled(BaseButton)`
  background-color: #FFFFFF;
  color: #2a2c2e;
`

/**
 * @version ./package.json
 */
const Button = ({ variant, dangerouslyAddClassName, ...rest }) => {
  const restNoDisabled = preventDisabling(rest)

  let ButtonComponent
  if (variant === 'secondary') {
    ButtonComponent = ButtonSecondary
  } else if (variant === 'inverted') {
    ButtonComponent = ButtonInverted
  } else {
    ButtonComponent = ButtonPrimary
  }
  return <ButtonComponent {...safeRest(restNoDisabled)} className={dangerouslyAddClassName} />
}

Button.propTypes = {
  /**
   * The HTML button type.
   */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  /**
   * The style.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted']),
  /**
   * The label. It can include the `A11yContent` component or strings.
   */
  children: or([PropTypes.string, componentWithName('A11yContent')]).isRequired,
}
Button.defaultProps = {
  type: 'button',
  variant: 'primary',
}

/** @component */
export default Button
