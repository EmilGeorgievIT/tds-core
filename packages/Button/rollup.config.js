import configure from '../../config/rollup.config'

export default configure({
  input: './Button.jsx',
})
