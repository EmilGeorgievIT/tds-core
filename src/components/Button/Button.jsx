import React from 'react'
import PropTypes from 'prop-types'

import { warn, deprecate } from '../../utils/warn'
import safeRest from '../../safeRest'

import styles from './Button.modules.scss'

const getClassName = (variant, invert) => {
  if (variant === 'primary' && invert) {
    warn('Button', 'Primary buttons cannot be inverted.')

    return styles.primary
  }

  if (invert) {
    return styles[`${variant}Inverted`]
  }

  return styles[variant]
}

const preventDisabling = ({ disabled, ...props }) => {
  if (disabled) {
    warn('Button', 'Buttons are not able to be disabled.')
  }

  return props
}

/**
 *
 * <span class="docs--badge green">updated!</span> <span class="docs--badge purple">v0.21.0</span>
 */
const Button = ({ type, variant, invert, children, ...rest }) => {
  const restNoDisabled = preventDisabling(rest)

  if (invert) {
    deprecate('Button', 'The invert prop is deprecated. Create an inverted Button with the inverted variant.')
  }

  if (variant === 'outlined') {
    deprecate('Button', 'The outlined variant is deprecated. Create an inverted Button with the inverted variant.')
  }

  return (
    <button {...safeRest(restNoDisabled)} type={type} className={getClassName(variant, invert)}>
      {children}
    </button>
  )
}

Button.propTypes = {
  /**
   * The HTML button type.
   */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  /**
   * The style.
   *
   * @since v0.21.0. Added 'inverted' to replace 'outlined'.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'outlined', 'inverted']),
  /**
   * Whether or not to invert the variant's color scheme.
   *
   * **Note:** Doesn't apply to `primary` variant.
   *
   * @deprecated since v0.21.0. Create inverted buttons with the inverted variant.
   */
  invert: PropTypes.bool,
  /**
   * The label.
   */
  children: PropTypes.string.isRequired
}
Button.defaultProps = {
  type: 'button',
  variant: 'primary',
  invert: false
}

export default Button
