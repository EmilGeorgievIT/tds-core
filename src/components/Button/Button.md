## Minimal Usage

Provide a function as the `onClick` prop to perform an action when clicked. **Avoid using a button if navigation 
is the primary action, as a link is more appropriate.**

### Recommendations

* Use buttons to move though a transaction.
* Aim to use only one button per page.
* Avoid excessively long button text.
* Make sure the button text describes an action.

By default, Buttons will be displayed in the `primary` variant. Use primary buttons for the main action on a page or 
in a form.

```
<Button>Submit</Button>
```

Specify the `variant` to create a button for secondary actions.

```
<Button variant="secondary">Find out more</Button>
```

## Placing buttons on solid colors

Use the `secondary` `invert` button on top of the solid TELUS purple.

```
const PurpleBlock = require('./__docs__/PurpleBlock').default;

<PurpleBlock>
  <Button variant="secondary" invert>Go back</Button>
</PurpleBlock>
```

## Placing buttons on images

Use the `outlined` variant when placing a button on top of an image, which will cause the image to show through. You can 
also `invert` the color scheme.

Use this variant with caution. There will be accessibility issues if the color contrast of the image and the button text is too low.

```
const Hero = require('./__docs__/Hero').default;

<Hero>
  <Button variant="outlined" invert>Go back</Button>
</Hero>
```
