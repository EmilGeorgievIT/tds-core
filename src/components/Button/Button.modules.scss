@import '../../scss/settings/colours';
@import '../../scss/settings/spacing';
@import '../../scss/settings/typography';
@import '../../scss/settings/responsive-grid';
@import '../../scss/utility/responsive';

$primary-bg-color: $color-primary;
$secondary-bg-color: $color-secondary;
$text-color: $color-white;


@mixin border {
  // Don't add a border because it changes the size of the button.
  // Box-shadow gives a "free" border using the text color.
  box-shadow: 0 0 0 1px;
}

// FIXME: Used for ButtonLink.
// Tried to do composes: primary from '../../Button/Button.modules.scss'; but unsuccessful due to rollup config + CSS Modules complications.
@mixin visited($text-color) {
  &:visited {
    color: $text-color;
  }
}

.button {
  @include helvetica-neue-medium-65;

  display: block;
  width: 100%;

  padding: $spacing-base $spacing-wide;

  border: 0;
  border-radius: 4px;

  transition: background .2s;

  font-family: $font-telus;
  font-size: 1rem;
  text-align: center;

  cursor: pointer;
}

@include from-breakpoint(medium) {
  .button {
    display: inline-block;
    width: auto;
    min-width: 180px;
  }

  .button + .button {
    // Using < 20px to compensate for the extra margin which appears between inline-block
    // elements due to whitespace in the actual markup.
    margin-left: 16px;
  }
}

.inverted {
  background-color: $text-color;

  &:hover {
    @include border;

    color: $text-color;
  }
}

.primary {
  composes: button;

  background-color: $primary-bg-color;
  color: $text-color;

  @include visited($text-color);

  &:hover {
    @include border;

    background-color: $text-color;
    color: $primary-bg-color;
  }
}

.secondary {
  composes: button;

  background-color: $secondary-bg-color;
  color: $text-color;

  @include visited($text-color);

  &:hover {
    @include border;

    background-color: $text-color;
    color: $secondary-bg-color;
  }
}

.secondaryInverted {
  composes: inverted button;

  color: $secondary-bg-color;

  @include visited($secondary-bg-color);

  &:hover {
    background-color: $secondary-bg-color;
    color: $text-color;
  }
}

.outlined {
  composes: button;

  @include border;

  background-color: transparent;
  color: $text-color;

  @include visited($text-color);

  &:hover {
    background-color: $text-color;
    color: $secondary-bg-color;

    box-shadow: none;
  }
}

.outlinedInverted {
  composes: inverted button;

  color: $secondary-bg-color;

  @include visited($secondary-bg-color);

  &:hover {
    @include border;

    background-color: transparent;
  }
}
