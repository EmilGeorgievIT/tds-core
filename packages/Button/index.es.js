import Button from './dist/index.es'

export default Button
