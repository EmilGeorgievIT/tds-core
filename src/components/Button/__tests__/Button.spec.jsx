import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { warn } from '../../../warn'

import Button from '../Button'

jest.mock('../../../warn', () => (
  { warn: jest.fn() }
))

describe('Button', () => {
  const doShallow = (overrides = {}) => shallow(<Button {...overrides}>Submit</Button>)

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const button = doShallow()

    expect(toJson(button)).toMatchSnapshot()
  })

  it('has one of the HTML button types', () => {
    let button = doShallow()
    expect(button).toHaveProp('type', 'button')

    button = doShallow({ type: 'reset' })
    expect(button).toHaveProp('type', 'reset')
  })

  it('can be presented as one of the allowed variants', () => {
    let button = doShallow()
    expect(button).toHaveClassName('primary')

    button = doShallow({ variant: 'primary' })
    expect(button).toHaveClassName('primary')

    button = doShallow({ variant: 'secondary' })
    expect(button).toHaveClassName('secondary')

    button = doShallow({ variant: 'outlined' })
    expect(button).toHaveClassName('outlined')
  })

  it('can be inverted outlined variant', () => {
    const outlinedButton = doShallow({ variant: 'outlined', invert: true })
    expect(outlinedButton).toHaveClassName('outlinedInverted')
  })

  it('can not be inverted for primary and secondary variants', () => {
    const primaryButton = doShallow({ variant: 'primary', invert: true })

    expect(primaryButton).toHaveClassName('primary')
    expect(warn).toHaveBeenCalled()

    const secondaryButton = doShallow({ variant: 'secondary', invert: true })

    expect(secondaryButton).toHaveClassName('secondary')
    expect(warn).toHaveBeenCalled()
  })

  it('can not be disabled', () => {
    const button = doShallow({ disabled: true })

    expect(button).not.toHaveProp('disabled')
    expect(warn).toHaveBeenCalled()
  })

  it('passes additional attributes to button element', () => {
    const button = doShallow({ id: 'the-button', tabindex: 1 })

    expect(button).toHaveProp('id', 'the-button')
    expect(button).toHaveProp('tabindex', 1)
  })

  it('does not allow custom CSS', () => {
    const button = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(button).not.toHaveProp('className', 'my-custom-class')
    expect(button).not.toHaveProp('style')
  })
})
