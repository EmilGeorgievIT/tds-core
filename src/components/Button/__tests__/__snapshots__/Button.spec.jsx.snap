// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<button
  className="button"
  type="button"
>
  <Flexbox
    dangerouslyAddClassName="primary"
    direction="row"
  >
    Submit
  </Flexbox>
</button>
`;
